### PR TITLE
Custom events added

### PR DESCRIPTION
**Pull request by iGEL**
As discussed in #136, there are cases when you want to log custom events (I want to log Delayed Job events). As far as I understand, it would be still possible to use just `Rails.logger.info` etc, but I think it's useful leverage the formatter settings. WDYT?

I've looked around and no class seemed to be the right place for this, so I've created a new module. Obviously I can move things around if you prefer something different.

I've changed one unrelated `send` call to `public_send`, since I consider `send` to be a smell.

### CODE REVIEW

**Comment by iGEL** [comment]
I've tried this in our app and realize, that for example the LogStash output contains the not as useful `message` field. Not sure, if the PR is still good, I went ahead and just log directly to Rails.logger without this.

File: lib/lograge/custom_event.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  module CustomEvent
+    class << self
+      def log(hash)
+        logger.public_send(Lograge.log_level, Lograge.formatter.call(hash))
+      end
+
+      private
+
+      def logger
+        Lograge.logger || defined?(Rails) && Rails.logger || fail('No logger available')
```
Comment (pxlpnk) [inline lib/lograge/custom_event.rb]: I think we should make this part a bit clearer.

File: lib/lograge/log_subscriber.rb [hunk 2]
```diff
@@ -18,7 +18,7 @@ def process_action(event)
 
       data = before_format(data, payload)
       formatted_message = Lograge.formatter.call(data)
-      logger.send(Lograge.log_level, formatted_message)
+      logger.public_send(Lograge.log_level, formatted_message)
```
Comment (pxlpnk) [inline lib/lograge/log_subscriber.rb]: :+1:

**Comment by pxlpnk** [comment]
Nice thank you so much!
I do not really understand yet why those builds fail. I try to investigate this later tonight when I have more headspace.

**Comment by pxlpnk** [comment]
Seems like a decency looks in logstash has breaking changes. This should make the tests work for now:
https://github.com/roidrage/lograge/pull/145

Could you add some documentation on how to use is in the readme?

What do you think @benlovell ?

**Comment by benlovell** [comment]
>I've tried this in our app and realize, that for example the LogStash output contains the not as useful >message field. Not sure, if the PR is still good, I went ahead and just log directly to Rails.logger without >this.

@iGEL could you expand on this?

**Comment by benlovell** [comment]
@iGEL Any interest in this?

**Comment by dsager** [comment]
This would actually be a a very useful feature. I hope it gets integrated at some point!

**Comment by pxlpnk** [comment]
@iGEL we would be happy to proceed with this. Could you maybe address the open points in the near future? :green_apple:

**Comment by highbass** [comment]
would love to get this merged with some documentation!

**Comment by benlovell** [comment]
@highbass Is this something you're using?

@iGEL Seems this one died from your perspective, I take it you went a different route?

**Comment by highbass** [comment]
i am not using it... but i was googling up on putting custom logs using lograge and came across this. I am currently playing around trying to see if it actually works but regardless would need it to be part of distribution or branch for me to be able to use it. Would be a very useful feature to have for custom logging with lograge for logstash.

**Comment by pxlpnk** [comment]
@highbass this shouldn't be to hard, you could use the default logstash formatter:  https://github.com/roidrage/lograge/blob/master/lib/lograge/formatters/logstash.rb
and make your own custom formatter: 
http://an-ti.eu/blog/ruby/2015/01/18/define-a-custom-lograge-formatter/

**Comment by santopuppy** [comment]
The CI failure seems to be related to this: https://github.com/rspec/rspec-mocks/issues/964. It seems the Jruby team has fixed this one.

**Comment by maniempire** [comment]
This feature is quite useful. I hope this PR will be merged soon.

**Comment by binarydev** [comment]
@benlovell can't see a way to rerun travis on this build, so i've forked master and have manually readded the changes made by @iGEL originally. Is there still any interest in this? Rspec test suite is passing and cleaned very minor issues for Rubocop identified via rake ci. If there's still interest, I'll submit a new MR with the every so slightly updated implementation.

**Comment by benlovell** [comment]
@binarydev Sure, it'll need a rebase and I'd like to see @pxlpnk's [comment addressed](https://github.com/roidrage/lograge/pull/144#r42516477) and a little documentation love - but definitely keen to get this in.

**Comment by binarydev** [comment]
@benlovell great! I've already put the latest changes on top of master locally. I'll just clarify the methods a bit and add some documentation. Quick question regarding the implementation: currently it creates a separate entry when using the JSON formatter rather then embedding it within a dedicated key for the request, which is great for DelayedJob, Sidekiq, etc. However, is this always ideal? If a custom event is fired off within a service class during a request, i would expect it to be considered part of the current request payload, perhaps under a new custom_events key as an array of events in chronological order.

I'm currently using lograge with the Elastic stack, and if I filter by requests in Kibana, all custom events would be excluded from the results under the current implementation based on my local testing. Should these entries instead be injected into the event payload for the request within which they are fired off? 

A solid example is a service class I have with a few submodules that act as a wrapper around R for stat analysis, where R has to execute a SQL query that Rails generates based on user selected criteria. Since Rails isn't executing the SQL, it won't be logged by lograge-sql, however I still want to track what SQL was generated for a request, along with the runtime performance data that we have R return to us as part of the results. It would be very strange for me to have this service return that data in the controller to be held in instance vars to provide #append_info_to_payload access to them.

**Comment by pxlpnk** [comment]
Hey @binarydev I recently ran into the same issue using lograge for sidekiq, it broke our logging as we had two separate JSON objects, instead of a unified payload. 
I personally think it would make sense to unify this. 
However I am not sure if I understand fully what you mean, would you mind posting some payload examples? I think this would help the discussion.

Looking forward to your PR! 🍏

**Comment by binarydev** [comment]
@pxlpnk Happy to share an example. So in the JSON log currently, as you said, a (formatted) request appears as 2 separate JSON objects. In the example below, I made a request, which calls a service, and within that service's submodules, I have the lines `Lograge::CustomEvent.log(my_special_event_message: "test1")` and `Lograge::CustomEvent.log(my_super_special_event_message: "test2")` separately being called, generating these 3 separate objects in the log:

```
{  
   "my_special_event_message":"test1",
   "@timestamp":"2017-05-04T15:44:46.905Z",
   "@version":"1",
   "message":"[]   (#)"
}
{  
   "my_super_special_event_message":"test2",
   "@timestamp":"2017-05-04T15:44:47.255Z",
   "@version":"1",
   "message":"[]   (#)"
}
{  
   "method":"GET",
   "path":"/",
   "format":"html",
   "controller":"HomeController",
   "action":"index",
   "status":200,
   "duration":177.21,
   "view":98.53,
   "db":54.1,
   "params":{},
   "user_id":1,
   "user_email":"admin@example.org",
   "user_role":"admin",
   "request_id":"b8e3b2a1-2f15-40eb-ac82-c19b9e2c2407",
   "@timestamp":"2017-05-04T15:44:47.305Z",
   "@version":"1",
   "message":"[200] GET / (HomeController#index)"
}
```

What I am proposing is injecting the custom events into the main request payload under a dedicated key as an error of triggered events, so that the above becomes the following
```
{  
   "method":"GET",
   "path":"/",
   "format":"html",
   "controller":"HomeController",
   "action":"index",
   "status":200,
   "duration":177.21,
   "view":98.53,
   "db":54.1,
   "params":{  

   },
   "user_id":1,
   "user_email":"admin@example.org",
   "user_role":"admin",
   "request_id":"b8e3b2a1-2f15-40eb-ac82-c19b9e2c2407",
   "@timestamp":"2017-05-04T15:44:47.305Z",
   "@version":"1",
   "message":"[200] GET / (HomeController#index)",
   "custom_events":[  
      {  
         "my_special_event_message":"test1",
         "@timestamp":"2017-05-04T15:44:46.905Z",
         "@version":"1",
         "message":"[]   (#)"
      },
      {  
         "my_super_special_event_message":"test2",
         "@timestamp":"2017-05-04T15:44:47.255Z",
         "@version":"1",
         "message":"[]   (#)"
      }
   ]
}
```

Where the custom_events keys holds the events triggered during the request cycle in chronological order, returning a single JSON object that can be consumed by logstash.

**Comment by pxlpnk** [comment]
I understand now. This is different from what I actually meant.
In my case sidekiq would issue a log line including two objects:
`{something: 'value'}, {something_else: 'another value}` 
I guess this is related to our implementation and a different problem.

@binarydev How would this work in the environment of something like sidekiq, where you have no access to the original performed request/logline?

@benlovell WDYT about the idea? I can see how this is useful, I am not sure though if this is a job that lograge should do. Or if it would make more sense to use the request ID to stitch the different logs in the logging tool.

**Comment by rolftimmermans** [comment]
Can this be merged please? Any sort of logging for custom events is better than none.

**Comment by lumir** [comment]
Thanks for this, I included that module in an initializer, and now I'm able to easily log custom events

**Pull request closed without merging**